src/common/hooks/useEndorsementChain/helpers.ts

```diff
@@ -0,0 +1,166 @@
+  let owner = "";
+  let holder = "";
+  for (const event of repeatedTransferEvents) {
+    owner = event.owner || owner;
+    holder = event.owner || holder;
+  }
+  return {
+    ...repeatedTransferEvents[0],
+    owner: owner,
+    holder: holder,
+  };
+};
+
+/*
+    Merge Transactions with the same transaction hash
+    E.g. 
+    TRANSFER_OWNERS that emits: both 
+    HOLDER_TRANSFER and OWNER_TRANSFER (on Title Escrow)
+    
+    INITIAL (Minting) that emits: 
+    (TRANSFER_OWNERS as above) and INITIAL (on Token Registry)
+    
+    SURRENDER_ACCEPTED (Shred) that emits: 
+    (TRANSFER_OWNERS as above) and SURRENDER_ACCEPTED (on Token Registry)
+*/
+export const mergeTransfers = (transferEvents: TransferBaseEvent[]): TransferBaseEvent[] => {
+  const hashList: Record<string, number[]> = {};
+  const repeatedHash: Set<string> = new Set();
+  for (let i = 0; i < transferEvents.length; i++) {
+    const transactionHash: string = transferEvents[i].transactionHash;
+    if (hashList[transactionHash] === undefined) {
+      hashList[transactionHash] = [i];
+    } else {
+      hashList[transactionHash].push(i);
+      repeatedHash.add(transactionHash);
+    }
+  }
+
+  const mergedEscrowTransfers: TransferBaseEvent[] = [];
+
+  for (const hash of Array.from(repeatedHash)) {
+    const length = hashList[hash].length;
+    const repeatedTransfers: TransferBaseEvent[] = [];
+    for (const index in hashList[hash]) {
+      repeatedTransfers.push(transferEvents[hashList[hash][index]]);
+    }
+    let type: TransferEventType = "INITIAL";
+    switch (length) {
+      case 2:
+        // 2 Repeated Transaction - Same Transaction Hash, Transaction Index && Block Number
+        // TitleEscrow: change_owner, change_holder
+        type = "TRANSFER_OWNERS";
+        break;
+      case 3:
+        // 3 Repeated Transaction - Same Transaction Hash, Transaction Index && Block Number
+        // Token-Registry: Initial
+        // TitleEscrow: change_owner, change_holder
+
+        // 3 Repeated Transaction - Same Transaction Hash, Transaction Index && Block Number
+        // Token-Registry: surrender_accepted - shred
+        // TitleEscrow: change_owner to 0x0, change_holder to 0x0
+        for (let i = 0; i < length; i++) {
+          if (repeatedTransfers[i].type === "INITIAL" || repeatedTransfers[i].type === "SURRENDER_ACCEPTED") {
+            type = repeatedTransfers[i].type;
+          }
+        }
+        break;
+      case 0:
+      case 1:
+      default:
+        throw new Error("Invalid repeated hash, update your configuration");
+    }
+    const mergedTransaction = mergeTransferTransactions(repeatedTransfers);
+    mergedEscrowTransfers.push({
+      ...mergedTransaction,
+      type,
+    } as TransferBaseEvent);
+  }
+  transferEvents = transferEvents.filter((x) => !repeatedHash.has(x.transactionHash));
+  return [...transferEvents, ...mergedEscrowTransfers];
+};
```

Comment:
the logic for mergeXfers can be refactored into somehting like this:

```
const grouped = groupBy(transferEvents, "transactionHash");
const vals = Object.values(grouped);
const res = vals.flatMap((group) => {
  if (group.length === 1) return group;
  if (group.length === 2) {
    const type = "TRANSFER_OWNERS" as TransferEventType;
    const base = group[0]
    const [owner, holder] = getHolderOwner(group);
    return [{ ...base, type, owner, holder }];
  }
  if (group.length === 3) {
    const base = group[0];
    let type = "INITIAL" as TransferEventType;
    const [owner, holder] = getHolderOwner(group);
    const found = group.find((x) => {
      return x.type === "INITIAL" || x.type === "SURRENDER_ACCEPTED";
    });
    return [{ ...base, owner, holder, type: found?.type || type }];
  }
});
```

test whether equivalent first, and add the corresponding comments and errors

Reply:
code for `getOwnerHolder`

```
const getHolderOwner = (events: TransferBaseEvent[]) => {
  let owner = "";
  let holder = "";
  for (const event of events) {
    owner = event.owner || owner;
    holder = event.owner || holder;
  }
  return [owner, holder];
};
```

Reply:
`groupBy` is taken from lodash